plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs

```diff
@@ -0,0 +1,124 @@
+import           Development.IDE.Core.Rules           (Log (LogShake))
+import           Development.IDE.Core.Shake           (addPersistentRule)
+import           Development.IDE.GHC.Compat
+import           Development.IDE.Types.HscEnvEq       (hscEnv)
+import           Ide.Plugin.Error                     (PluginError (PluginInternalError, PluginRuleFailed),
+                                                       getNormalizedFilePathE)
+import           Ide.Plugin.SemanticTokens.Mappings
+import           Ide.Plugin.SemanticTokens.Query
+import           Ide.Plugin.SemanticTokens.Types
+import           Ide.Types
+import qualified Language.LSP.Protocol.Lens           as L
+import           Language.LSP.Protocol.Message        (Method (Method_TextDocumentSemanticTokensFull))
+import           Language.LSP.Protocol.Types          (NormalizedFilePath,
+                                                       SemanticTokens,
+                                                       type (|?) (InL))
+import           Prelude                              hiding (span)
+
+logWith :: (MonadIO m) => IdeState -> Priority -> String -> m ()
+logWith st prior = liftIO . logPriority (ideLogger st) prior . T.pack
+
+-----------------------
+---- the api
+-----------------------
+
+computeSemanticTokens :: IdeState -> NormalizedFilePath -> ExceptT PluginError Action SemanticTokens
+computeSemanticTokens state nfp = do
+    let dbg = logWith state Debug
+    dbg $ "Computing semantic tokens for: " <> show nfp
+    (RangeHsSemanticTokenTypes {tokens}, mapping) <- useWithStaleE GetSemanticTokens nfp
+    let rangeTokens = mapMaybe (\(span, name) -> (,name) <$> toCurrentRange mapping span) tokens
+    withExceptT PluginInternalError $ liftEither $ semanticTokenAbsoluteSemanticTokens rangeTokens
+
+semanticTokensFull :: PluginMethodHandler IdeState 'Method_TextDocumentSemanticTokensFull
+semanticTokensFull state _ param = do
+  nfp <- getNormalizedFilePathE (param ^. L.textDocument . L.uri)
+  items <- runActionE "SemanticTokens.semanticTokensFull" state $ computeSemanticTokens state nfp
+  return $ InL items
+
+-- | Defines the 'getSemanticTokensRule' function, compute semantic tokens for a Haskell source file.
+--
+-- This Rule collects information from various sources, including:
+--
+-- Imported name token type from Rule 'GetSemanticTokens'
+-- Local names token type from 'hieAst'
+-- Name locations from 'hieAst'
+-- Visible names from 'tmrRenamed'
+--
+-- It then combines this information to compute the semantic tokens for the file.
+--
+getSemanticTokensRule :: Recorder (WithPriority Log) -> Rules ()
+getSemanticTokensRule recorder =
+  define (cmapWithPrio LogShake recorder) $ \GetSemanticTokens nfp -> do
+    (TcModuleResult {..}) <- use_ TypeCheck nfp
+    (hscEnv -> hsc) <- use_ GhcSessionDeps nfp
+    (HAR {..}) <- use_ GetHieAst nfp
+    Just (_, ast) <- return $ listToMaybe $ Map.toList $ getAsts hieAst
```

Comment:
partial match? this will throw if it fails

Reply:
I want to let the rule fail if ast is not here, not sure what is better practice🤔

Reply:
Sure, but this will fail with a pattern match exception which I don't think will be caught. Not 100% sure. Better to fail explicitly, there are some functions for doing this with things returning Maybe somewehre.

Reply:
Maybe we return Nothing in the result?

Reply:
See e.g. `handleMaybe`

Reply:
Thanks I will try this out.

Reply:
done